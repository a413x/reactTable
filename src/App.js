import {useState} from 'react'
import {Table} from './components/Table.js'
import {LoadData} from './components/LoadData.js'
import {sortData} from './utils/utils.js'
import './styles/App.css';

const mock_data = [
  {
    id : 1,
    firstName : 'A',
    lastName: 'B',
    email: 'example@example.com',
    phone: '12345678'
  },
  {
    id : 2,
    firstName : 'C',
    lastName: 'D',
    email: 'example@example1.com',
    phone: '123456789'
  }
]

const App = () => {
  const [data, setData] = useState(mock_data)

  const sortCallback = (colObj, order) => {
    const sortedData = sortData(data, colObj.name, order)
    setData(sortedData)
  }

  const loadCallback = (data) => {
    setData(data)
  }

  return (
    <div className="app">
      <LoadData loadCallback = {loadCallback} />
      <Table data = {data} sortCallback = {sortCallback}/>
    </div>
  );
}

export default App
