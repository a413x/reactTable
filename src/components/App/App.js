import {useState} from 'react'
import {Table} from '../Table/Table.js'
import {LoadData} from '../LoadData/LoadData.js'
import {AddForm} from '../AddForm/AddForm.js'
import {PagesNav} from '../PagesNav/PagesNav.js'
import {SearchForm} from '../SearchForm/SearchForm.js'
import {Details} from '../Details/Details.js'
import {sortData, searchData} from '../../utils/utils.js'
import {mock_data} from '../../__mock_data__/mock_data.js'

let storedData = [...mock_data]

const App = () => {
  const [data, setData] = useState([...storedData])
  const [page, setPage] = useState(1)
  const [rowsOnPage, setRowsOnPage] = useState(5)
  const [selectedData, setSelectedData] = useState(null)
  const [sortedCol, setSortedCol] = useState({name: '', order: ''})

  const sortCallback = (colObj, order, newSortedCol) => {
    const sortedData = sortData(data, colObj.name, order)
    storedData = sortData(storedData, colObj.name, order)
    setData(sortedData)
    setSortedCol({name: colObj.name, order: order})
  }

  const searchCallback = search => {
    const filtered = searchData(storedData, search)
    if(filtered.length === storedData.length) {
      setData(storedData)
    }
    setPage(1)
    setData(filtered)
  }

  const loadCallback = data => {
    storedData = [...data]
    setSortedCol({name: '', order: ''})
    setData(data)
  }

  const addCallback = dataObj => {
    const newData = [...data]
    newData.push(dataObj)
    setData(newData)
  }

  const pagesCount = Math.ceil(data.length/rowsOnPage)
  const pageChangeCallback = newPage => {
    if(newPage < 1 || newPage > pagesCount) return
    setPage(newPage)
  }

  const rowsNumChangeCallback = newRowsNum => {
    setPage(1)
    setRowsOnPage(newRowsNum)
  }

  const showDetailsCallback = dataObj => setSelectedData({...dataObj})
  const closeDetailsCallback = () => setSelectedData(null)

  const from = (page-1)*rowsOnPage
  const dataToShow = data.slice(from, from + rowsOnPage)

  return (
    <div className = 'app container'>
      <LoadData loadCallback = {loadCallback} />
      <AddForm addCallback = {addCallback} />
      <SearchForm searchCallback = {searchCallback} />
      <Table
        data = {dataToShow}
        sortedCol = {sortedCol}
        sortCallback = {sortCallback}
        showDetailsCallback = {showDetailsCallback}
      />
      <PagesNav
        page = {page} rowsOnPage = {rowsOnPage}
        pagesCount = {pagesCount}
        pageChangeCallback = {pageChangeCallback}
        rowsNumChangeCallback = {rowsNumChangeCallback}
      />
      {selectedData &&
        <Details
          dataObj = {selectedData}
          closeDetailsCallback = {closeDetailsCallback}
        />
      }
    </div>
  )
}

export default App
