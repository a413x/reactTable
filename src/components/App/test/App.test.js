import {render, fireEvent, waitFor} from '@testing-library/react'
import App from '../App.js'

describe('App component tests', () => {
  test('renders properly', () => {
    const {container} = render(<App />)
    expect(container).toMatchSnapshot()
  })
})
