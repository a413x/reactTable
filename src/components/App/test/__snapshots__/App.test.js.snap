// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests renders properly 1`] = `
<div>
  <div
    class="app container"
  >
    <div
      class="load-data mt-3"
    >
      <span
        class="mr-3"
      >
        Load data:
      </span>
      <button
        class="btn btn-primary mr-3"
      >
        Small
      </button>
      <button
        class="btn btn-primary"
      >
        Big
      </button>
    </div>
    <div
      class="add-form"
    >
      <button
        class="btn btn-info"
      >
        Add
      </button>
    </div>
    <div
      class="search-form my-3"
    >
      <div
        class="form-inline"
      >
        <label>
          Search:
        </label>
        <input
          class="form-control mx-2"
          type="search"
        />
      </div>
    </div>
    <div
      class="table"
    >
      <table
        class="table table-hover"
      >
        <thead>
          <tr>
            <th
              class=""
            >
              #
            </th>
            <th
              class=""
            >
              First name
            </th>
            <th
              class=""
            >
              Last name
            </th>
            <th
              class=""
            >
              E-mail
            </th>
            <th
              class=""
            >
              Phone
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              Frederick
            </td>
            <td>
              Catania
            </td>
            <td>
              GHellems@magna.net
            </td>
            <td>
              (458)969-6847
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              Curtis
            </td>
            <td>
              Bias
            </td>
            <td>
              CRobichaud@nec.gov
            </td>
            <td>
              (481)075-5034
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              Dora
            </td>
            <td>
              Crenshaw
            </td>
            <td>
              MSanchez@at.org
            </td>
            <td>
              (959)257-7541
            </td>
          </tr>
          <tr>
            <td>
              4
            </td>
            <td>
              Grace
            </td>
            <td>
              French
            </td>
            <td>
              JLund@vestibulum.com
            </td>
            <td>
              (976)555-7134
            </td>
          </tr>
          <tr>
            <td>
              5
            </td>
            <td>
              Randy
            </td>
            <td>
              Bunch
            </td>
            <td>
              KGunter@lacus.io
            </td>
            <td>
              (941)539-5519
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="pages-nav d-flex"
    >
      <button
        class="btn btn-light"
        disabled=""
      >
        ◀
      </button>
      <nav
        class="d-flex w-100 justify-content-around"
      >
        <div
          class="form-inline"
        >
          <label>
            Page
          </label>
          <input
            class="form-control m-2"
            style="width: 100px;"
            type="number"
            value="1"
          />
          <label>
            of 
            1
          </label>
        </div>
        <div
          class="form-inline"
        >
          <label
            class="mr-2"
          >
            Rows on page:
          </label>
          <select>
            <option>
              5
            </option>
            <option>
              10
            </option>
            <option>
              50
            </option>
          </select>
        </div>
      </nav>
      <button
        class="btn btn-light"
        disabled=""
      >
        ▶
      </button>
    </div>
  </div>
</div>
`;
